File: src/main/golang/design/PoisonPill.go
package design

import (
	"context"
	"log"
)

type PoisonPill[T comparable] struct {
	Name     string
	Queue    chan T
	Consumer func(T)
	Poison   T
	ctx      context.Context
	cancel   context.CancelFunc
}

func NewPoisonPill[T comparable](name string, poison T, queueSize int, consumer func(T)) *PoisonPill[T] {
	ctx, cancel := context.WithCancel(context.Background())
	return &PoisonPill[T]{
		Name:     name,
		Queue:    make(chan T, queueSize),
		Consumer: consumer,
		Poison:   poison,
		ctx:      ctx,
		cancel:   cancel,
	}
}

func (p *PoisonPill[T]) Offer(item T) {
	p.Queue <- item
}

func (p *PoisonPill[T]) Start() {
	log.Printf("%s ready to run", p.Name)
	go func() {
		defer log.Printf("%s close", p.Name)
		for {
			select {
			case <-p.ctx.Done():
				return
			case item := <-p.Queue:
				if item == p.Poison {
					log.Printf("%s received poison pill, exiting", p.Name)
					p.cancel()
					return
				}
				p.Consumer(item)
			}
		}
	}()
}

func (p *PoisonPill[T]) Shutdown() {
	p.cancel()
}

File: src/main/py/design/PoisonPill.py
import threading
import logging
from typing import TypeVar, Generic, Callable
from queue import Queue, Empty

T = TypeVar('T')
log = logging.getLogger(__name__)

class PoisonPill(Generic[T], threading.Thread):
    def __init__(self, name: str, poison: T, queue: Queue, consumer: Callable[[T], None]):
        super().__init__()
        self.name = name.upper()
        self.queue = queue
        self.consumer = consumer
        self.poison = poison
        self._shutdown = threading.Event()

    def offer(self, item: T):
        self.queue.put(item)

    def before(self):
        logger.info(f"{self.name} ready to run")

    def after(self):
        logger.info(f"{self.name} close")

    def run(self):
        self.before()
        while not self._shutdown.is_set():
            try:
                item = self.queue.get(timeout=1)  # Avoid infinite block
                if item == self.poison:
                    logger.info(f"{self.name} received poison pill, exiting")
                    self._shutdown.set()
                    break
                self.consumer(item)
            except Empty:
                continue
        self.after()

    def shutdown(self):
        self._shutdown.set()